File: src/lib/balance.ts
import type { AdhocTransaction, DayBalance, RecurringTransaction, SkippedOccurrence, TxEntry } from '../types'
import { expandRecurring } from './recurrence'

export function localDateStr(d = new Date()): string {
  return `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, '0')}-${String(d.getDate()).padStart(2, '0')}`
}

function buildTxMap(
  recurring: RecurringTransaction[],
  adhoc: AdhocTransaction[],
  skipped: SkippedOccurrence[],
  fromDate: string,
  toDate: string
): Map<string, { deposits: TxEntry[]; expenses: TxEntry[] }> {
  const map = new Map<string, { deposits: TxEntry[]; expenses: TxEntry[] }>()

  const ensure = (date: string) => {
    if (!map.has(date)) map.set(date, { deposits: [], expenses: [] })
    return map.get(date)!
  }

  const skipMap = new Map<string, string>()
  for (const s of skipped) {
    skipMap.set(`${s.transaction_id}|${s.date}`, s.id)
  }

  const [fromY, fromM] = fromDate.split('-').map(Number)
  const [toY, toM] = toDate.split('-').map(Number)

  let y = fromY, m = fromM
  while (y < toY || (y === toY && m <= toM)) {
    for (const rule of recurring) {
      if (!rule.active) continue
      for (const date of expandRecurring(rule, y, m)) {
        if (date < fromDate || date > toDate) continue
        const skippedId = skipMap.get(`${rule.id}|${date}`) ?? null
        const day = ensure(date)
        const entry: TxEntry = { id: rule.id, name: rule.name, amount: rule.amount, source: 'recurring', skipped: skippedId !== null, skippedId }
        if (rule.type === 'deposit') day.deposits.push(entry)
        else day.expenses.push(entry)
      }
    }
    m++
    if (m > 12) { m = 1; y++ }
  }

  for (const tx of adhoc) {
    if (tx.date < fromDate || tx.date > toDate) continue
    const skippedId = skipMap.get(`${tx.id}|${tx.date}`) ?? null
    const day = ensure(tx.date)
    const entry: TxEntry = { id: tx.id, name: tx.name, amount: tx.amount, source: 'adhoc', skipped: skippedId !== null, skippedId }
    if (tx.type === 'deposit') day.deposits.push(entry)
    else day.expenses.push(entry)
  }

  return map
}

/**
 * startBalance = actual bank balance as of startDate (reflects only cleared/completed transactions).
 * cutoffDate   = first day of first month the tool was used; transactions before this are ignored.
 *
 * Past days (before today): show transaction badges but no balance number.
 * Today and future: project balance forward using all uncompleted transactions.
 * The adjustment accounts for uncompleted transactions between cutoffDate and startDate-1
 * so they are factored into today's opening balance.
 */
export function computeAllDailyBalances(
  startBalance: number,
  startDate: string,
  recurring: RecurringTransaction[],
  adhoc: AdhocTransaction[],
  skipped: SkippedOccurrence[],
  cutoffDate: string,
  fromDate: string,
  toDate: string
): Record<string, DayBalance> {
  const todayStr = localDateStr()

  // Build tx map only from cutoffDate onwards — pre-cutoff transactions are ignored
  const txMap = buildTxMap(recurring, adhoc, skipped, cutoffDate, toDate)

  const allDates: string[] = []
  const cur = new Date(fromDate + 'T00:00:00')
  const end = new Date(toDate + 'T00:00:00')
  while (cur <= end) {
    allDates.push(cur.toISOString().slice(0, 10))
    cur.setDate(cur.getDate() + 1)
  }

  const result: Record<string, DayBalance> = {}

  // Adjustment: sum uncompleted transactions from cutoffDate up to (but not including) startDate.
  // These haven't cleared the bank yet, so they aren't in startBalance.
  let adjustment = 0
  for (const date of allDates) {
    if (date < cutoffDate || date >= startDate) continue
    const { deposits, expenses } = txMap.get(date) ?? { deposits: [], expenses: [] }
    adjustment += deposits.filter(t => !t.skipped).reduce((s, t) => s + t.amount, 0)
    adjustment -= expenses.filter(t => !t.skipped).reduce((s, t) => s + t.amount, 0)
  }

  // Forward pass from startDate: project balance using uncompleted transactions.
  // Days before today get endBalance: null (display badges only, no balance shown).
  let fwdBal = startBalance + adjustment
  for (const date of allDates) {
    if (date < startDate) continue
    const { deposits, expenses } = txMap.get(date) ?? { deposits: [], expenses: [] }
    const net = deposits.filter(t => !t.skipped).reduce((s, t) => s + t.amount, 0)
             - expenses.filter(t => !t.skipped).reduce((s, t) => s + t.amount, 0)
    fwdBal += net
    const isPast = date < todayStr
    result[date] = {
      deposits,
      expenses,
      endBalance: isPast ? null : fwdBal,
      isToday: date === todayStr,
      isPast,
    }
  }

  // Past days from cutoffDate to startDate-1: show badges but no balance
  for (const date of allDates) {
    if (date < cutoffDate || date >= startDate) continue
    const { deposits, expenses } = txMap.get(date) ?? { deposits: [], expenses: [] }
    result[date] = { deposits, expenses, endBalance: null, isToday: false, isPast: true }
  }

  return result
}

File: ios/BudgetBuddy/Logic/Balance.swift
import Foundation

struct TxEntry: Identifiable, Hashable {
    enum Source: String, Hashable { case recurring, adhoc }
    let id: String              // composite: "<txId>|<date>" so SwiftUI lists are stable across days
    let txId: String
    let name: String
    let amount: Double          // always positive — sign comes from `type`
    let type: TransactionType
    let source: Source
    var skipped: Bool
    var skippedId: String?

    /// Negative for expenses, positive for deposits.
    var signedAmount: Double { type == .expense ? -amount : amount }
}

struct DayBalance: Hashable {
    var deposits: [TxEntry] = []
    var expenses: [TxEntry] = []
    var endBalance: Double?     // nil = past day, no projection shown
    var isToday: Bool = false
    var isPast: Bool = false
}

enum Balance {
    /// Mirrors src/lib/balance.ts `computeAllDailyBalances`. Returns a map keyed by `YYYY-MM-DD`.
    static func computeDaily(
        startBalance: Double,
        startDate: String,
        recurring: [RecurringTransaction],
        adhoc: [AdhocTransaction],
        skipped: [SkippedOccurrence],
        cutoffDate: String,
        fromDate: String,
        toDate: String,
        today: String = Calendar.todayYMD()
    ) -> [String: DayBalance] {
        let txMap = buildTxMap(recurring: recurring, adhoc: adhoc, skipped: skipped,
                               fromDate: cutoffDate, toDate: toDate)
        let allDates = Self.dateRange(from: fromDate, to: toDate)
        var result: [String: DayBalance] = [:]

        // Adjustment: uncompleted transactions between cutoff and start that
        // aren't yet reflected in the user-entered startBalance.
        var adjustment = 0.0
        for date in allDates where date >= cutoffDate && date < startDate {
            let day = txMap[date] ?? DayBalance()
            adjustment += day.deposits.filter { !$0.skipped }.reduce(0) { $0 + $1.amount }
            adjustment -= day.expenses.filter { !$0.skipped }.reduce(0) { $0 + $1.amount }
        }

        var fwdBal = startBalance + adjustment
        for date in allDates where date >= startDate {
            let day = txMap[date] ?? DayBalance()
            let net = day.deposits.filter { !$0.skipped }.reduce(0) { $0 + $1.amount }
                    - day.expenses.filter { !$0.skipped }.reduce(0) { $0 + $1.amount }
            fwdBal += net
            let isPast = date < today
            result[date] = DayBalance(
                deposits: day.deposits,
                expenses: day.expenses,
                endBalance: isPast ? nil : fwdBal,
                isToday: date == today,
                isPast: isPast
            )
        }
        for date in allDates where date >= cutoffDate && date < startDate {
            let day = txMap[date] ?? DayBalance()
            result[date] = DayBalance(deposits: day.deposits, expenses: day.expenses,
                                      endBalance: nil, isToday: false, isPast: true)
        }
        return result
    }

    private static func buildTxMap(
        recurring: [RecurringTransaction],
        adhoc: [AdhocTransaction],
        skipped: [SkippedOccurrence],
        fromDate: String,
        toDate: String
    ) -> [String: DayBalance] {
        var map: [String: DayBalance] = [:]
        var skipMap: [String: String] = [:]   // "<txId>|<date>" → skippedId
        for s in skipped { skipMap["\(s.transaction_id)|\(s.date)"] = s.id }

        let fromParts = fromDate.split(separator: "-").compactMap { Int($0) }
        let toParts = toDate.split(separator: "-").compactMap { Int($0) }
        guard fromParts.count == 3, toParts.count == 3 else { return map }
        var (y, m) = (fromParts[0], fromParts[1])
        let (toY, toM) = (toParts[0], toParts[1])

        while y < toY || (y == toY && m <= toM) {
            for rule in recurring where rule.active != 0 {
                for date in Recurrence.expand(rule, year: y, month: m) {
                    if date < fromDate || date > toDate { continue }
                    let skippedId = skipMap["\(rule.id)|\(date)"]
                    let entry = TxEntry(
                        id: "\(rule.id)|\(date)",
                        txId: rule.id, name: rule.name, amount: rule.amount,
                        type: rule.type, source: .recurring,
                        skipped: skippedId != nil, skippedId: skippedId
                    )
                    var day = map[date] ?? DayBalance()
                    if rule.type == .deposit { day.deposits.append(entry) } else { day.expenses.append(entry) }
                    map[date] = day
                }
            }
            m += 1
            if m > 12 { m = 1; y += 1 }
        }

        for tx in adhoc {
            if tx.date < fromDate || tx.date > toDate { continue }
            let skippedId = skipMap["\(tx.id)|\(tx.date)"]
            let entry = TxEntry(
                id: "\(tx.id)|\(tx.date)",
                txId: tx.id, name: tx.name, amount: tx.amount,
                type: tx.type, source: .adhoc,
                skipped: skippedId != nil, skippedId: skippedId
            )
            var day = map[tx.date] ?? DayBalance()
            if tx.type == .deposit { day.deposits.append(entry) } else { day.expenses.append(entry) }
            map[tx.date] = day
        }

        return map
    }

    /// Inclusive YYYY-MM-DD list from `from` to `to`.
    static func dateRange(from: String, to: String) -> [String] {
        guard let start = Calendar.parseYMD(from),
              let end = Calendar.parseYMD(to), start <= end else { return [] }
        var dates: [String] = []
        var cur = start
        let cal = Calendar.gregorian
        while cur <= end {
            dates.append(Calendar.ymdString(cur))
            guard let next = cal.date(byAdding: .day, value: 1, to: cur) else { break }
            cur = next
        }
        return dates
    }
}
